tests/multigrid/dof_06.cc

```diff
@@ -0,0 +1,172 @@
+#include <deal.II/base/exceptions.h>
+
+#include <deal.II/dofs/dof_handler.h>
+
+#include <deal.II/grid/grid_generator.h>
+#include <deal.II/grid/grid_tools.h>
+#include <deal.II/grid/tria.h>
+
+#include <deal.II/multigrid/mg_constrained_dofs.h>
+
+#include <vector>
+
+constexpr int DIM = 2;
+
+using namespace dealii;
+
+template <typename FaceType>
+void
+print_face(const FaceType &face, unsigned int level)
+{
+  AssertThrow(
+    face->get_fe(0).n_dofs_per_face() > 0,
+    ExcMessage(
+      "Please use a finite element with at least 1 DoF for this test"));
+
+  std::vector<types::global_dof_index> mg_dofs(
+    face->get_fe(0).n_dofs_per_face());
+  face->get_mg_dof_indices(level, mg_dofs);
+
+  // Print face DoFs
+  std::cout << mg_dofs[0];
+  for (unsigned int i_dof = 1; i_dof < mg_dofs.size(); ++i_dof)
+    std::cout << ", " << mg_dofs[i_dof];
+}
+
+template <int dim>
+void
+check()
+{
+  typedef Triangulation<dim> Tria;
+  // Create triangulation
+  Tria tria(Tria::limit_level_difference_at_vertices);
+
+  // Create coarse grid
+  GridGenerator::hyper_cube(tria, -1, 1, true);
+
+  // Create periodic boundary along x axis
+  std::vector<GridTools::PeriodicFacePair<typename Tria::cell_iterator>>
+    face_pairs;
+
+  GridTools::collect_periodic_faces(tria, 0, 1, 0, face_pairs);
+
+  tria.add_periodicity(face_pairs);
+
+  // Refine once and then refine the corner again
+  tria.refine_global(1);
+
+  tria.begin_active()->set_refine_flag();
+  tria.prepare_coarsening_and_refinement();
+  tria.execute_coarsening_and_refinement();
+
+  // Assemble DoFs
+  DoFHandler<dim> dof_handler(tria);
+  FE_Q<dim>       fe_q(1);
+
+  dof_handler.distribute_dofs(fe_q);
+  dof_handler.distribute_mg_dofs();
+
+  // Try to assemble MGConstrainedDofs
+  MGConstrainedDoFs constrained_dofs;
+  constrained_dofs.initialize(dof_handler);
+
+  // Report which DIM is ok
+  std::cout << "--- DIM " << dim << " OK ---" << std::endl;
+
+  // Debug printouts:
+  for (unsigned int level = 0; level < tria.n_global_levels(); ++level)
+    {
+      std::cout << "Level " << level << ": " << std::endl;
+
+      // Print faces that connect to coarser DoF
+      std::cout << "  Faces neighboring coarser cells:" << std::endl;
+      for (typename DoFHandler<dim>::cell_iterator level_cell =
+             dof_handler.begin(level);
+           level_cell != dof_handler.end(level);
+           ++level_cell)
+        {
+          // Iterate over all faces
+          for (unsigned int i_face = 0; i_face < 2 * dim; ++i_face)
```

Comment:
We prefer `GeometryInfo<dim>::faces_per_cell` over the explicit `2*dim`. Would you mind changing that here and below?

Reply:
Done ( unless I missed one).